util/activation_memory_profiler.py

```diff
@@ -0,0 +1,134 @@
+            size = num_bytes_from_shape_and_dtype(tensor_spec.shape, tensor_spec.dtype)
+            stack_trace = node.meta.get("stack_trace")
+            fqn = _get_module_hierarchy(node)
+            for j in range(start, end + 1):
+                if memory_timeline[j] is None:
+                    memory_timeline[j] = MemoryTimeline()
+                memory_timeline[j].allocations.append(
+                    Allocation(node.name, node.target, i, size, fqn, stack_trace)
+                )
+    return memory_timeline
+
+
+def _validate_memory_planning_is_done(exported_program: ExportedProgram):
+    """
+    Validate whether the memory planning has been done on the given program.
+    """
+    for node in exported_program.graph.nodes:
+        # If there is at least one memory allocation node, then we know the memory planning has been done.
+        if node.target == memory.alloc:
+            return True
+    return False
+
+
+def generate_memory_trace(
+    executorch_program_manager: ExecutorchProgramManager, chrome_trace_filename: str
+):
+    """
+    Generate the memory timeline from the given Executorch program.
+    Args:
+        executorch_program The Executorch program to be analyzed.
+    Returns:
+        Chrome trace in JSON format:
+        Format:
+        Each thread represents a unit of time. Thus to navigate timeline scroll up and down.
+        For each thread, the x axis represents live tensor objects that are normalized according the allocation size.
+    """
+    if not isinstance(executorch_program_manager, ExecutorchProgramManager):
+        raise ValueError(
+            f"generate_memory_trace expects ExecutorchProgramManager instance but got {type(executorch_program_manager)}"
+        )
+
+    exported_program = executorch_program_manager.exported_program()
+    if not _validate_memory_planning_is_done(exported_program):
+        raise ValueError("Executorch program does not have memory planning.")
+
+    memory_timeline = create_tensor_allocation_info(exported_program.graph)
+    root = {}
+    trace_events = []
+    root["traceEvents"] = trace_events
+
+    tid = 0
+    for i, memory_timeline_event in enumerate(memory_timeline):
+        start_time = 0
+        if memory_timeline_event is None:
+            continue
+        for allocation in memory_timeline_event.allocations:
+            e = {}
+            e["name"] = allocation.name
+            e["cat"] = "memory_allocation"
+            e["ph"] = "X"
+            e["ts"] = int(start_time)
+            allocation_size_kb = allocation.size_bytes
+            e["dur"] = int(allocation_size_kb)
+            e["pid"] = 0
+            e["tid"] = tid
+            e["args"] = {}
+            e["args"]["op_name"] = f"{allocation.op_name}"
+            e["args"]["fqn"] = f"{allocation.fqn}"
+            e["args"]["source"] = f"{allocation.file_and_line_num}"
+            e["args"]["bytes"] = allocation.size_bytes
+            start_time += allocation_size_kb
```

Comment:
refactor into a fn?

Reply:
Sure. Let me do in future refactor (hopefully not something that is just kicked down the road)